File: array/238.py
#O(n^2), TLE
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        ret = list(itertools.repeat(1, len(nums)))
        
        for i in range(len(nums)):
            for j in range(len(ret)):
                if i != j:
                    ret[j] = ret[j]*nums[i]
                
        return ret
            
# O(n)
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        ret = []
        temp = 1
        for i in range(0, len(nums)):
            ret.append(temp)
            temp *= nums[i]
        
        temp = 1
        for i in range(len(nums)-1, -1, -1):
            ret[i] *= temp
            temp *= nums[i]
        
        return ret
            
        

File: array/238.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ret = new int[nums.length];
        ret[0] = 1;
        
        for (int i = 1; i < nums.length; i++) {
            ret[i] = nums[i-1]*ret[i-1];
        }
        int accu = 1;
        for (int i = nums.length-1; i >= 0; i--) {
            ret[i] *= accu;
            accu *= nums[i];
        }
        return ret;
    }
}
